src/Controller/PostfinanceResponseController.php

```diff
@@ -34,17 +37,26 @@ public function processAcceptResponse(Request $request, PaymentInterface $paymen
     $plugin_definition = $payment->getPaymentMethod()->getPluginDefinition();
 
     // Generate local SHASign.
-    $request_values = $request->query->all();
+    $request_values = simplexml_load_string($request->query->get('data'));
 
-    // Generate SHA-OUT signature
-    $sha_sign = PostfinanceHelper::generateShaSign($request_values, $plugin_definition['sha_out_key']);
+    $attribute_array = array();
+    $attributes = $request_values->attributes();
 
-    if ($sha_sign == $request->get('SHASIGN')) {
+    foreach ($attributes as $key => $val) {
+      $attribute_array[(string) $key] = (string) $val;
+    }
+    $sha_sent = array_pop($attribute_array);
+
+    // Generate SHA-OUT signature.
+    $sha_sign = PostfinanceHelper::generateShaSign($attribute_array, $plugin_definition['sha_out_key']);
+
+    if ($sha_sign == $sha_sent) {
       drupal_set_message(t('Payment succesfull.'), 'error');
       $this->savePayment($payment, 'payment_success');
-    } else {
+    }
+    else {
       $this->savePayment($payment, 'payment_failed');
```

Comment:
The save and return doesn't look correct yet. We need to assert that we are redirect to the correct page in the test. We somehow change the return vlaue in savePayment(), but we're not using the return value here, or in any of the other methods here.